File: src/pipeline/fix.ts
import type { ApplyFixFn } from "./types";
import type { ReviewIssue, SceneObject } from "../scene/schema";

// Applies review fixes to a scene, producing the next-pass SceneState. Works on
// a deep clone so the caller's scene is never mutated, then bumps `pass`. Each
// issue's Fix is applied in order; later fixes simply compound on earlier ones.
export const fix: ApplyFixFn = (scene, issues) => {
  const next = structuredClone(scene);
  for (const issue of issues) {
    const obj = next.objects.find((candidate) => candidate.id === issue.objectId);
    if (!obj) {
      throw new Error(`fix: issue targets unknown object "${issue.objectId}"`);
    }
    applyFix(obj, issue);
  }
  next.pass = scene.pass + 1;
  return next;
};

function applyFix(obj: SceneObject, issue: ReviewIssue): void {
  const { fix: f } = issue;
  switch (f.op) {
    case "move": {
      if (!f.delta) {
        throw new Error(`fix: "move" on "${issue.objectId}" is missing delta`);
      }
      const [px, py, pz] = obj.transform.position;
      obj.transform.position = [px + f.delta[0], py + f.delta[1], pz + f.delta[2]];
      return;
    }
    case "rotate": {
      if (f.rotationYDeg === undefined) {
        throw new Error(`fix: "rotate" on "${issue.objectId}" is missing rotationYDeg`);
      }
      obj.transform.rotationYDeg += f.rotationYDeg;
      return;
    }
    case "resize": {
      if (f.scaleFactor === undefined) {
        throw new Error(`fix: "resize" on "${issue.objectId}" is missing scaleFactor`);
      }
      obj.transform.scale *= f.scaleFactor;
      // Re-seat on the floor at the new scale. `position` is the object CENTER
      // (scene/schema.ts) and a floor-resting center sits at half the object's scaled
      // height — exactly geometryCheck's restY = approxSize·scale/2 (which layout.ts also
      // emits at scale 1). Both renderers place the pivot at position.y and scale the model
      // (base seated at pivot-local -approxSize[1]/2) about it, so the rendered base lands at
      // position.y - scale·approxSize[1]/2. Without this update the stale position.y no longer
      // matches the new scale, so the base sinks (scaleFactor>1) or floats (scaleFactor<1) — in
      // both the live viewer and the headless critic's own render, so the critic would judge a
      // mis-seated scene. Keeping the invariant here rests it for every downstream reader.
      obj.transform.position[1] = (obj.approxSize[1] * obj.transform.scale) / 2;
      return;
    }
    case "regenerate": {
      if (f.newMeshyPrompt) {
        obj.meshyPrompt = f.newMeshyPrompt;
      }
      obj.glbUrl = undefined;
      obj.status = "pending";
      return;
    }
  }
}

File: backend/src/pipeline/fix.py
"""Apply review fixes to a scene, producing the next-pass SceneState.

Mirrors src/pipeline/fix.ts. Works on a deep clone so the caller's scene is
never mutated, then bumps `pass`.
"""

from scene.schema import Fix, ReviewIssue, SceneObject, SceneState


def fix(scene: SceneState, issues: list[ReviewIssue]) -> SceneState:
    """Apply fixes to a deep copy of the scene and return the next pass state."""
    next_state = scene.model_copy(deep=True)
    for issue in issues:
        obj = _find_object(next_state, issue.object_id)
        if obj is None:
            raise RuntimeError(f"fix: issue targets unknown object '{issue.object_id}'")
        _apply_fix(obj, issue)
    next_state.pass_ = scene.pass_ + 1
    return next_state


def _find_object(state: SceneState, object_id: str) -> SceneObject | None:
    for obj in state.objects:
        if obj.id == object_id:
            return obj
    return None


def _apply_fix(obj: SceneObject, issue: ReviewIssue) -> None:
    f: Fix = issue.fix
    op = f.op
    if op == "move":
        if f.delta is None:
            raise RuntimeError(f'fix: "move" on "{issue.object_id}" is missing delta')
        px, py, pz = obj.transform.position
        obj.transform.position = (
            px + f.delta[0],
            py + f.delta[1],
            pz + f.delta[2],
        )
    elif op == "rotate":
        if f.rotation_y_deg is None:
            raise RuntimeError(
                f'fix: "rotate" on "{issue.object_id}" is missing rotationYDeg'
            )
        obj.transform.rotation_y_deg += f.rotation_y_deg
    elif op == "resize":
        if f.scale_factor is None:
            raise RuntimeError(
                f'fix: "resize" on "{issue.object_id}" is missing scaleFactor'
            )
        obj.transform.scale *= f.scale_factor
        # Re-seat on the floor at the new scale.
        obj.transform.position = (
            obj.transform.position[0],
            (obj.approx_size[1] * obj.transform.scale) / 2,
            obj.transform.position[2],
        )
    elif op == "regenerate":
        if f.new_meshy_prompt is not None:
            obj.meshy_prompt = f.new_meshy_prompt
        obj.glb_url = None
        obj.status = "pending"
    else:
        raise RuntimeError(f'fix: unknown op "{op}" on "{issue.object_id}"')
